conans/test/functional/toolchain/test_msbuild.py

```diff
@@ -230,17 +307,25 @@ class WinTest(unittest.TestCase):
         class App(ConanFile):
             settings = "os", "arch", "compiler", "build_type"
             requires = "hello/0.1"
-            generators = "msbuild"
             options = {"shared": [True, False]}
             default_options = {"shared": False}
             def toolchain(self):
                 tc = MSBuildToolchain(self)
+                if self.options["hello"].shared and self.settings.build_type == "Release":
+                    tc.configuration = "ReleaseShared"
+                    tc.generator.configuration = "ReleaseShared"
```

Comment:
A Question: the generator configuration should follow always the configuration of the project or not necessarily? Should be the same by default or not really?
In other words, if the generator configuration is different, will the dependencies will be found for the configuration specified in the project?

Reply:
Yes, after conversation with users, it is very typical that the configuration of the project refers to the way it is linking dependencies. It could be both, of course, but users can do:

- ReleaseStatic: I am building my app executable, linking it statically with its dependencies
- ReleaseShared: My app executable links with shared libs.

While other users or other projects could define:

- ReleaseDLL: I am building a shared library myself
- ReleaseStatic: I am building a static library myself

So it seems that the system should be flexible to let the users define the logic they want differently to generators and toolchains.

Reply:
Mhhh, so answering my questions...
- The `tc.configuration` and `tc.generator.configuration` might be different.
- We don't want to default the `tc.generator.configuration` with the value adjusted in the `tc.configuration`.

Did you mean the matrix is the following?:

- ReleaseStatic (`tc.configuration`): I am building my app executable, linking it statically with its dependencies
- ReleaseShared (`tc.generator.configuration`) : My app executable links with shared libs.
  
While other users or other projects could define:

- ReleaseDLL (`tc.configuration`): I am building a shared library myself
 - ReleaseStatic (`tc.generator.configuration`): I am building a static library myself

Reply:
I think in the second case, it would be both ``tc.configuration``, assuming that dependencies in both cases do not change at all. 

In the first case, it seems that ``tc.generator.configuration`` => "tc.configuration", in most cases both should be used, as it sounds difficult to consume that specific dependencies configuration unless your current project doesn't define that configuration. Though it might be possible to use the same toolchain file for different configurations, so maybe not in all cases.

Reply:
I so sorry but I don't get it hahahah. Could you do a Matrix of possible values of `tc.configuration` and `tc.generator.configuration` and the meaning?

Reply:
The name of the config is irrelevant, it is just something a user will define. So lets define it from the other side:

- ``toolchain()`` defines ``tc.configuration`` only. This means that generators will keep generating the same standard files, like ``zlib_release.cmake`` or ``zlib_release_x86_v140.props``. The dependencies will be exactly the same, using standard Conan ``build_type`` settings. 
  This would be the case if the current project have several custom configurations in cmake or Visual for building different variants, like building a shared library or a static library. The dependencies are the same (static, shared, or whatever), they do not change, but the current project build from the IDE can be different.

- ``toolchain()`` defines ``generator.configuration`` only. This means that dependencies change based on some inputs, for example if the user wants to be creating an executable linked with static deps or with shared deps. The current toolchain files could be identical, and do not depend on the configuration, like building an executable, maybe the toolchain doesn't change at all: same generator, same architecture, etc.

- ``toolchain()`` defines both ``generator.configuration`` and ``tc.configuration``: Both the dependencies and the current toolchain will change based on the current configuration. Like if the previous case, but building that executable against shared or static libraries would require some extra flags to be defined by the toolchain, e.g. passing ``-PIE`` or some LinkTimeOptimization flag when building against static or shared libraries.

Please let me know if this makes sense.